schema/scope_test.go

```diff
@@ -313,3 +313,78 @@ func TestSelfSerialization(t *testing.T) {
 		t.Fatalf("Unexpected root object: %s", serializedScopeMap["root"])
 	}
 }
+
+//nolint:funlen
+func TestApplyingExternalNamespace(t *testing.T) {
+	// This test tests applying with scopes, properties, and objects, maps, and lists.
+	// It must be passed through all of those types. Validating this makes sure that
+	// it gets applied all the way down, and errors are propagated up.
+	var testScope = schema.NewScopeSchema(
+		schema.NewObjectSchema(
+			"scopeTestObjectA",
+			map[string]*schema.PropertySchema{
+				"b": schema.NewPropertySchema(
+					schema.NewNamespacedRefSchema("scopeTestObjectB", "test-namespace", nil),
+					nil,
+					true,
+					nil,
+					nil,
+					nil,
+					nil,
+					nil,
+				),
+				"list-type": schema.NewPropertySchema(
+					schema.NewListSchema(
+						schema.NewNamespacedRefSchema("scopeTestObjectB", "test-namespace", nil),
+						nil,
+						nil,
+					),
+					nil,
+					true,
+					nil,
+					nil,
+					nil,
+					nil,
+					nil,
+				),
+				"map-type": schema.NewPropertySchema(
+					schema.NewMapSchema(
+						schema.NewIntSchema(nil, nil, nil),
+						schema.NewNamespacedRefSchema("scopeTestObjectB", "test-namespace", nil),
+						nil,
+						nil,
+					),
+					nil,
+					true,
+					nil,
+					nil,
+					nil,
+					nil,
+					nil,
+				),
+			},
+		),
+	)
+	var externalScope = schema.NewScopeSchema(
+		schema.NewObjectSchema(
+			"scopeTestObjectB",
+			map[string]*schema.PropertySchema{
+				"c": schema.NewPropertySchema(
+					schema.NewStringSchema(nil, nil, nil),
+					nil,
+					true,
+					nil,
+					nil,
+					nil,
+					nil,
+					nil,
+				),
+			},
+		),
+	)
+	// Not applied yet
+	assert.Error(t, testScope.ValidateReferences())
+	testScope.ApplyScope(externalScope, "test-namespace")
+	// Now it's applied, so the error should be resolved.
+	assert.NoError(t, testScope.ValidateReferences())
```

Comment:
It gets applied to a scope, which then gets applied to every property inside of it, which includes a ref schema, a list of ref schemas, and a map with ref schema values. If it was not applied to all of them, there would be an error.

Is there a comment that you think would clarify this?

Reply:
> If it was not applied to all of them, there would be an error.

Really?  What if one of the mid-level structures returned "early" instead of recursing on its members?  How would we know?

> Is there a comment that you think would clarify this?

No...I think we should have _assertions_ which check the bottom-most schemas, independent of the top-level one, to see that the scope was applied.  I suggest putting some entries in your list and map and then checking that calling `ValidateReferences()` on _them_ works.  E.g., map values might be either lists or maps; lists can contain lists or maps; check a few of the bottom-level schemas (as it stands, "all the way down" is not very far...).

Reply:
I split it up to use a scope for each case in a table format. That should address the pitfalls of the prior test case. And I modified the comment.